### PR TITLE
Mixed precision topography_mod

### PR DESCRIPTION
**Pull request by mcallic2**
**Description**
Adds mixed precision capabilities to topography_mod

Fixes # (issue)

**How Has This Been Tested?**
Please describe the tests that you ran to verify your changes. Please also note
any relevant details for your test configuration (e.g. compiler, OS).  Include
enough information so someone can reproduce your tests.

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: topography/include/topography.inc [hunk 1]
```diff
@@ -703,272 +443,152 @@ end interface
...
+subroutine FIND_INDICES_( ybeg, yend, ydat, js, je )
+real(kind=FMS_TOP_KIND_), intent(in)  :: ybeg, yend, ydat(:)
+integer,                  intent(out) :: js, je
+integer                               :: j
+integer, parameter                    :: lkind = FMS_TOP_KIND_
 
    js = 1
    do j = 1, size(ydat(:))-1
-      if (ybeg >= ydat(j) .and. ybeg <= ydat(j+1)) then
+      if (ybeg >= ydat(j) .and. ybeg <= ydat(real(j,FMS_TOP_KIND_)+1.0_lkind)) then
```
Comment (mlee03) [inline topography/include/topography.inc:468]: I don't think this needs a real, I think j+1 is the index of ydat

File: topography/include/topography.inc [hunk 2]
```diff
@@ -703,272 +443,152 @@ end interface
...
          js = j
          exit
       endif
    enddo
 
    je = size(ydat(:))-1
    do j = js, size(ydat(:))-1
-      if (yend >= ydat(j) .and. yend <= ydat(j+1)) then
+      if (yend >= ydat(j) .and. yend <= ydat(real(j,FMS_TOP_KIND_)+1.0_lkind)) then
```
Comment (mlee03) [inline topography/include/topography.inc:476]: here too

**Review by mlee03 — CHANGES_REQUESTED** [review]

File: topography/include/topography.inc (hunk 2 above)
Comment (mcallic2) [reply topography/include/topography.inc]: resolved in b7400af4

File: topography/include/topography.inc (hunk 1 above)
Comment (mcallic2) [reply topography/include/topography.inc]: resolved in b7400af4

File: topography/gaussian_topog.F90 [hunk 3]
```diff
@@ -105,9 +110,9 @@ module gaussian_topog_mod
 !! by variables in the namelist.
 subroutine gaussian_topog_init ( lon, lat, zsurf )
```
Comment (mlee03) [inline topography/gaussian_topog.F90:111]: Is there a reason why this subroutine wasn't moved to the inc file?

File: topography/include/gaussian_topog.inc [hunk 4]
```diff
@@ -162,112 +59,53 @@ end subroutine gaussian_topog_init
...
+  dtr = tpi/360.0_lkind
 
 ! defaults and convert degrees to radians (dtr)
-  olon = 90.*dtr;  if (present(olond)) olon=olond*dtr
-  olat = 45.*dtr;  if (present(olatd)) olat=olatd*dtr
-  wlon = 15.*dtr;  if (present(wlond)) wlon=wlond*dtr
-  wlat = 15.*dtr;  if (present(wlatd)) wlat=wlatd*dtr
-  rlon =  0.    ;  if (present(rlond)) rlon=rlond*dtr
-  rlat =  0.    ;  if (present(rlatd)) rlat=rlatd*dtr
+  olon = 90.0_lkind*dtr;  if (present(olond)) olon=olond*dtr
```
Comment (mlee03) [inline topography/include/gaussian_topog.inc:89]: olon=real(olond*dtr, lkind) and for below too

File: topography/include/gaussian_topog.inc [hunk 5]
```diff
@@ -162,112 +59,53 @@ end subroutine gaussian_topog_init
...
+  wlon = 15.0_lkind*dtr;  if (present(wlond)) wlon=wlond*dtr
+  wlat = 15.0_lkind*dtr;  if (present(wlatd)) wlat=wlatd*dtr
+  rlon = 0.0_lkind     ;  if (present(rlond)) rlon=rlond*dtr
+  rlat = 0.0_lkind     ;  if (present(rlatd)) rlat=rlatd*dtr
 
 ! compute gaussian-shaped mountain
     do j=1,size(lat(:))
       dy = abs(lat(j) - olat)   ! dist from y origin
-      yy = max(0., dy-rlat)/wlat
+      yy = max(0.0_lkind, dy-rlat)/wlat
```
Comment (mlee03) [inline topography/include/gaussian_topog.inc:99]: all module levels variables are declared as r8_kind.  you need to convert them to lkind

**Review by mlee03 — CHANGES_REQUESTED** [review]

File: topography/topography.F90 [hunk 6]
```diff
@@ -224,651 +253,115 @@ module topography_mod
...
 !##################   private interfaces below here   ##################
 !#######################################################################
 
- function open_topog_file ( )
- logical :: open_topog_file
- real    :: r_ipts, r_jpts
- integer :: namelen
+function open_topog_file ( )
+logical :: open_topog_file
+real(kind=r8_kind)    :: r_ipts, r_jpts
```
Comment (mlee03) [inline topography/topography.F90:287]: Im wondering if these could be r4_kind if the whole purpose of this subroutine is to open the topog_file and get ipts and jpts

**Comment by mlee03** [comment]
@mcallic2 i made a PR to your branch

File: topography/include/gaussian_topog.inc [hunk 7]
```diff
@@ -123,151 +88,63 @@ integer :: n
...
+  dtr = tpi/360.0_lkind
 
 ! defaults and convert degrees to radians (dtr)
-  olon = 90.*dtr;  if (present(olond)) olon=olond*dtr
-  olat = 45.*dtr;  if (present(olatd)) olat=olatd*dtr
-  wlon = 15.*dtr;  if (present(wlond)) wlon=wlond*dtr
-  wlat = 15.*dtr;  if (present(wlatd)) wlat=wlatd*dtr
-  rlon =  0.    ;  if (present(rlond)) rlon=rlond*dtr
-  rlat =  0.    ;  if (present(rlatd)) rlat=rlatd*dtr
+  olon = 90.0_lkind*dtr;  if (present(olond)) olon=real(olond*dtr,r8_kind)
```
Comment (mlee03) [inline topography/include/gaussian_topog.inc:128]: sorry missed this one, olon= 90.0_r8_kind, and for the rest below too

File: topography/include/gaussian_topog.inc [hunk 8]
```diff
@@ -29,72 +29,36 @@
...
-!   <DATA NAME="rlon, rlat" UNITS="degree" TYPE="real" DIM="(mxmtns)" DEFAULT="0.">
-!     The longitude and latitude half-width of mountain ridges (in degrees).  For a
-!     "standard" Gaussian mountain set rlon=rlat=0.
-!    </DATA>
-!
-!    <DATA NAME="NOTE">
-!     The variables in this namelist are only used when routine
-!     <TT>gaussian_topog_init</TT> is called.  The namelist variables
-!     are dimensioned (by 10), so that multiple mountains can be generated.
+!> @brief Returns a simple surface height field that consists of a single
```
Comment (rem1776) [inline topography/include/gaussian_topog.inc:35]: can you move this comment block to right above the function its documenting (get_gaussian_topog). If its not right above the routine it won't get parsed correctly.

**Review by rem1776 — CHANGES_REQUESTED** [review]

File: topography/include/gaussian_topog.inc (hunk 7 above)
Comment (mcallic2) [reply topography/include/gaussian_topog.inc]: fixed in 7be92cd7

File: topography/include/gaussian_topog.inc (hunk 8 above)
Comment (mcallic2) [reply topography/include/gaussian_topog.inc]: fixed in b55e5dad

File: topography/include/topography.inc [hunk 9]
```diff
@@ -738,237 +479,116 @@ end interface
...
-                 ocean(im,jm) .or. ocean(ip,jm) .or. ocean(ip,jp) .or. ocean(im,jp)) then
-                 ocean(i,j) = .true.
-                 new = new + 1
-             endif
-          endif
-       enddo
-       enddo
+         do i = 1, m
+            if (.not.ocean(i,j) .and. pctwater(i,j) > ocean_pct_crit) then
+                 im = i-1; ip = i+1; jm = j-1; jp = j+1
```
[the remark this answers is not in the document]
Comment (mcallic2) [reply topography/include/topography.inc]: fixed in d9ed1452

File: topography/include/gaussian_topog.inc [hunk 10]
```diff
@@ -29,72 +29,34 @@
...
-!   <DATA NAME="rlon, rlat" UNITS="degree" TYPE="real" DIM="(mxmtns)" DEFAULT="0.">
-!     The longitude and latitude half-width of mountain ridges (in degrees).  For a
-!     "standard" Gaussian mountain set rlon=rlat=0.
-!    </DATA>
-!
-!    <DATA NAME="NOTE">
-!     The variables in this namelist are only used when routine
-!     <TT>gaussian_topog_init</TT> is called.  The namelist variables
-!     are dimensioned (by 10), so that multiple mountains can be generated.
+!> The height, position, width, and elongation of the mountain
```
Comment (mlee03) [inline topography/include/gaussian_topog.inc:36]: I think this comment block has been misplaced?

Comment (mcallic2) [reply topography/include/gaussian_topog.inc:36]: @rem1776 told me to move it for documentation, it's right above `get_gaussian_topog` now

Comment (mlee03) [reply topography/include/gaussian_topog.inc:36]: it seems to be above gaussian_topog_init?

File: topography/include/gaussian_topog.inc (hunk 8 above)
Comment (rem1776) [reply topography/include/gaussian_topog.inc]: You only moved the first 3 lines (the brief description), the rest ofit needs to be moved (lines 35-59)

File: topography/include/gaussian_topog.inc (hunk 10 above)
Comment (rem1776) [reply topography/include/gaussian_topog.inc:36]: https://github.com/NOAA-GFDL/FMS/pull/1250#discussion_r1310661855

File: topography/include/gaussian_topog.inc (hunk 8 above)
Comment (mlee03) [reply topography/include/gaussian_topog.inc]: @mcallic2 could you make these changes?

Comment (mlee03) [reply topography/include/gaussian_topog.inc]: I'm still lost as to why the comments for get_gaussian_topog is moved here above gaussian_topog_init...

**Comment by mcallic2** [comment]
@mlee03 @rem1776 friends, I went back and read through the comments. I understand what both of you are saying.  I "reorganized" the comment blocks (in be19229c) to coincide with their respective routines. Let me know if ya'll agree.

File: topography/include/gaussian_topog.inc (hunk 8 above)
Comment (mlee03) [reply topography/include/gaussian_topog.inc:34]: Do lines 34-35 also need to be moved down?

Comment (rem1776) [reply topography/include/gaussian_topog.inc:34]: I think they are fine, it'll just read the first !> as the brief descriptions and the second part as the expanded.

**Review by rem1776 — APPROVED** [review]

File: topography/include/gaussian_topog.inc [hunk 11]
```diff
@@ -123,10 +123,10 @@ integer, parameter     :: lkind = FMS_TOP_KIND_
   dtr = tpi/360.0_lkind
 
 ! defaults and convert degrees to radians (dtr)
-  olon = 90.0_r8_kind*dtr;  if (present(olond)) olon=real(olond*dtr,r8_kind)
-  olat = 45.0_r8_kind*dtr;  if (present(olatd)) olat=real(olatd*dtr,r8_kind)
-  wlon = 15.0_r8_kind*dtr;  if (present(wlond)) wlon=real(wlond*dtr,r8_kind)
-  wlat = 15.0_r8_kind*dtr;  if (present(wlatd)) wlat=real(wlatd*dtr,r8_kind)
+  olon = 90.0_r8_kind*real(dtr, r8_kind);  if (present(olond)) olon=real(olond*dtr,r8_kind)
```
Comment (mlee03) [inline topography/include/gaussian_topog.inc:126]: real(dtr,r8_kind) isn't necessary but ok.

**Pull request merged**